webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeConfigurationService.scala

```diff
@@ -0,0 +1,350 @@
+    ZIO.foreach_(invalidatedIds.map(_._2) : List[CacheComplianceQueueAction])(action =>
+    {
+      (for {
+        _  <- performAction(action)
+      } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating NodeConfiguration cache for node: [${action.nodeId.value}]: ${err.fullMsg}"))
+    }
+    )
+  )
+  // start updating
+  updateCacheFromRequest.forever.forkDaemon.runNow
+  /**
+   * Clear cache. Try a reload asynchronously, disregarding
+   * the result
+   */
+  override def clearCache(): Unit = {
+    init().runNow
+    logEffect.debug("Node expected reports cache cleared")
+  }
+
+
+  /**
+   * Do something with the action we received
+   */
+  private[this] def performAction(action: CacheComplianceQueueAction): IOResult[Unit] = {
+    import CacheComplianceQueueAction._
+    // in a semaphore
+    semaphore.withPermit(
+       action match {
+                          case insert: InsertNodeInCache => IOResult.effectNonBlocking {  cache = cache + (insert.nodeId -> None) }
+                          case delete: RemoveNodeInCache => IOResult.effectNonBlocking {  cache = cache.removed(delete.nodeId) }
+                          case update: UpdateNodeConfiguration => IOResult.effectNonBlocking {  cache = cache + (update.nodeId -> Some(update.nodeConfiguration)) }
+                          case something =>
+                            Inconsistency(s"NodeConfiguration service cache received unknown command : ${something}").fail
+                           // should not happen
+                   }
+    )
+  }
+
+  /**
+   * invalidate with an action to do something
+   * order is important
+   */
+  def invalidateWithAction(actions: Seq[(NodeId, CacheComplianceQueueAction)]): IOResult[Unit] = {
+    ZIO.when(actions.nonEmpty) {
+      ReportLoggerPure.Cache.debug(s"Node Configuration cache: invalidation request for nodes with action: [${actions.map(_._1).map { _.value }.mkString(",")}]") *>
+        invalidateMergeUpdateSemaphore.withPermit(for {
+          elements     <- invalidateNodeConfigurationRequest.takeAll
+          allActions   =  (elements.flatten ++ actions)
+          _            <- invalidateNodeConfigurationRequest.offer(allActions)
+        } yield ())
+    }
+  }
+  // ? question ?
+  // how to properly ensure that cache is synchro ?
+  // we have the begin date of the nodeexpectedreport that my offer a way to ensure that we don't replace
+  // a value with something older
+
+  // Merge the data with the cache
+  // Strategy:
+  // * if nodeId is not in cache, remove the nodeId
+  // * if the nodeId is in cache, and value is None, augment cache
+  // * if the nodeid is in cache, and value is not None, pick the most recent value to augment cache
+  // Returns the values merged with cache
+  private[this] def mergeDataFromDBWithCache(fromDb: Map[NodeId, Option[NodeExpectedReports]]):IOResult[ Map[NodeId, Option[NodeExpectedReports]] ] = {
```

Comment:
I'm not sure that that function should be split appart from the "getCurrentExpectedReports" part. It makes harder to see what need to be atomic, and it hides the fact that nodes "fromDb" should all have beed not initialized in cache (since it's called for that reason). 
Let me try to merge them

Reply:
i *think* it needs to be appart. It'll be probably used in other methods

Reply:
but for now, it is not, and it hads substential complexity.